From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Tue, 2 Sep 2025 13:55:30 -0500
Subject: [PATCH] Fish Parallel World Ticking API

This patch provides an API for performance monitoring plugins like PurpurBars.
A better approach would be to include an event when the world is done ticking,
  which will allow to re-use the RollingAverage logic from Spark and Minecraft internals.
However, since every fork developer will want to have said event into their own namespace,
  it will be virtually impossible to provide a universal API.
With this approach only new methods are added into already existing Bukkit API, and the same
  naming conventions as Bukkit were used, which means there's a bigger chance of this API to
  get standarized.

diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0ec885b97d655c03371840d739b47063494ed36f..c9034f4520daf0603137f9fbd654b94d186d0c6e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2761,4 +2761,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
+
+    // Fish start - Parallel World Ticking API
+    /**
+     * Returns whether the software has the Parallel World Ticking feature enabled
+     *
+     * @return If the Parallel World Ticking feature is enabled
+     */
+    public default boolean isParallelWorldTickingEnabled() {
+        return false;
+    }
+    // Fish end - Parallel World Ticking API
 }
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index b202bdb84bae23f7b2b5af591879a032f6378a25..e2d8dc16480cdca7b679d6c105ad787885a08bd9 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4560,6 +4560,25 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @NotNull
     public Collection<GeneratedStructure> getStructures(int x, int z, @NotNull Structure structure);
 
+    // Fish start - Parallel World Ticking API
+    /**
+     * Get a sample of the world last tick times (in nanos)
+     *
+     * @return A sample of the world last tick times (in nanos)
+     */
+    public default long @NotNull [] getTickTimes() {
+        return new long[0];
+    }
+    /**
+     * Get the world's average tick time (in millis)
+     *
+     * @return World's local average tick time (in millis)
+     */
+    public default double getAverageTickTime () {
+        return 0.0;
+    }
+    // Fish end - Parallel World Ticking API
+
     /**
      * Represents various map environment types that a world may be
      */
